### PR TITLE
executor,planner: use an object pool to reuse PlanBuilder

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve?

Problem Summary:

Reduce object allocation for https://github.com/pingcap/tidb/issues/25573

### What is changed and how it works?

What's Changed:

Just like the parser object, now the PlanBuilder object are reused.

How it Works:

```
func() {
	builder := planBuilderPool.Get().(*plannercore.PlanBuilder)
	defer planBuilderPool.Put(builder.ResetForReuse())
	builder.Init(sctx, is, hintProcessor)
        ...
}
```

The PlanBuilder is take from a pool, and after build finish, it's put back to the pool.

### Check List 

Tests 

- [ ] No code

```
cd session
go test -test.run TestXXX -test.bench BenchmarkPreparedPointGet -memprofile mem.out -benchmem -benchtime 30s
```

```
Before:
BenchmarkPreparedPointGet-16    	 4129156	      8657 ns/op	    7054 B/op	      85 allocs/op

After:
BenchmarkPreparedPointGet-16    	 4254156	      8388 ns/op	    6509 B/op	      81 allocs/op
```

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Some exported API changes.

- plancore.NewPlanBuilder()
- (*PlanBuilder) Init()
- (*PlanBuilder) ResetForReuse()

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note 

- No release note

### CODE REVIEW

**Review by eurekaka — APPROVED** [review]

File: planner/core/planbuilder.go [hunk 1]
```diff
@@ -601,25 +607,66 @@ func (b *PlanBuilder) popSelectOffset() {
...
+}
+
+// ResetForReuse reset the plan builder, put it into pool for reuse.
+// After reset for reuse, the object should be equal to a object returned by NewPlanBuilder().
+func (b *PlanBuilder) ResetForReuse() *PlanBuilder {
+	// Save some fields for reuse.
+	saveOuterCTEs := b.outerCTEs[:0]
+	saveColMapper := b.colMapper
+	for k := range saveColMapper {
+		delete(saveColMapper, k)
```
Comment (mmyj) [inline planner/core/planbuilder.go:647]: Why we need to delete those keys here? Is there any difference  between deleting keys and remake a new map in this case?

File: planner/core/planbuilder.go [hunk 2]
```diff
@@ -513,6 +513,12 @@ type handleColHelper struct {
 	stackTail         int
 }
 
+func (hch *handleColHelper) resetForReuse() {
+	*hch = handleColHelper{
+		id2HandleMapStack: hch.id2HandleMapStack[:0],
+	}
+}
```
Comment (mmyj) [inline planner/core/planbuilder.go:520]:
```
hch.id2HandleMapStack = hch.id2HandleMapStack[:0]
hch.stackTail = 0
```
Could we reset `hch.id2HandleMapStack` directly?

File: planner/core/planbuilder.go (hunk 1 above)
Comment (tiancaiamao) [reply planner/core/planbuilder.go:647]: Allocation is expensive, especially for a GC language like Go.

```
for k := range m {
        delete(m, k)
}
```

is optimized to a function `runtime.mapclear()`,  while make(map[]) is also a function call.
The they're both one function call, the different is, one spent time on allocation, the other spent time on clear up the map.

So here is a tiny benchmark to compare:

```
package main

import (
	"testing"
)

func BenchmarkReuse(b *testing.B) {
	m := make(map[int]interface{})
	b.ResetTimer()
	for i:=0; i<b.N; i++ {
		use(m)
		for k := range m {
			delete(m, k)
		}
	}
}

func BenchmarkDrop(b *testing.B) {
	var m map[int]interface{}
	b.ResetTimer()
	for i:=0; i<b.N; i++ {
		m = make(map[int]interface{})
		use(m)
	}
}

func use(m map[int]interface{}) {
	m[32] = "23"
	m[46] = "421"
	m[23] = "423"
	m[52] = "666"
	m[52] = "54"
}
```

The result is:

```
genius@genius-System-Product-Name:~/project/sss$ go test -run XXX  -bench Benchmark -benchmem
goos: linux
goarch: amd64
pkg: sss
cpu: AMD Ryzen 7 PRO 4750G with Radeon Graphics
BenchmarkReuse-16    	16455187	        72.58 ns/op	       0 B/op	       0 allocs/op
BenchmarkDrop-16     	 6761996	       178.3 ns/op	     256 B/op	       2 allocs/op
PASS
ok  	sss	2.657s
```

As you can see the result ---- "allocation is expensive"

File: planner/core/planbuilder.go (hunk 2 above)
Comment (tiancaiamao) [reply planner/core/planbuilder.go:520]: I do this deliberately.

```
*hch = handleColHelper{
		id2HandleMapStack: hch.id2HandleMapStack[:0],
}
```


When someone update the `handleColHelper` struct and add some more fields,
he may not notice the `resetForReuse` function, and forget to update it. 
Then things goes wrong.

While in my coding style, I mean reuse the `hch.id2HandleMapStack` field and only that field, reset other fields to the default value.

For example, when field `xxx string` is added, 

```
type handleColHelper struct {
    id2HandleMapStack []map[int64][]HandleCols
    stackTail int
    xxx string
}
```

Forget to update `resetForReuse` is not a serious mistake... the **code is robust**!

File: planner/core/planbuilder.go (hunk 1 above)
Comment (mmyj) [reply planner/core/planbuilder.go:647]: Thanks!

**Review by mmyj — APPROVED** [review]
lgtm

**Pull request merged**